Inline colors in TextEdit

poke1024 opened the pull request:
Prologue:

The current master has a cool function to open a color picker in the code editor using the context menu and `Pick Color`, but it's well hidden and I wasn't aware of it until I found it in the code:

<img width="312" alt="master-picker" src="https://user-images.githubusercontent.com/11859538/34070082-5216f076-e25f-11e7-9805-46e17d59728d.png">

This PR removes this great but hidden context menu `Pick Color` in favor of inline colors, i.e. colored circles in front of `Color` (you can click on them to open the color picker):

<img width="529" alt="inline-color" src="https://user-images.githubusercontent.com/11859538/34070090-5d3740dc-e25f-11e7-9d30-51206d4ba772.png">

Full demo: https://www.dropbox.com/s/fdzqwnt6sm77w3u/demo-inline-colors.mp4?dl=0

I'm currently using unicode `0x25CF` to draw the color circle, I'm not sure if that's safe on all versions of Freetype/Linux (plus sides: it adapts to all font sizes and is much faster than `draw_circle`).

l1ebl committed:
Inline colors in TextEdit

Hinsbart commented:
Nice feature :)
However, I don't think it's wise to put this code in `TextEdit`.

groud commented:
Nice! Great idea.

volzhs commented:
I agree with @Hinsbart

brainsick commented:
I'm not qualified to review the code, but a couple questions come to mind.  Does this have any implication for code that cares about a line's length?  I'm thinking of the recent patches for code indentation behavior that impacted the selection area.

This updates ClassDB entries, specifically moves _color_changed from ScriptTextEditor to TextEdit.  Is this then considered part of the public API?  Should that be documented?  Not familiar with community guidelines on things that start with _; _color_changed doesn't appear to be documented now, so it's at least status quo.  :)

poke1024 commented:
It's indeed an API change so it's Godot 3.1.

The reason for the signal move is that `TextEdit` draws and handles the color circle (and knows when it's clicked). We could still move the color picker handling to `ScriptTextEditor` (probably a good idea?), but `TextEdit` would be the one to send the signal "show it now".

There should be no impact on selection code or the like as `TextEdit` collates the `C` character of `Color` with the color circle as if both were one glyph at rendering level. Any code above does not see the circle in positions or selection areas or the like - it's happening only inside the rendering and does not take up a position.

That's also the reason I believe this thing belongs in `TextEdit` and not somewhere else; that's the safest way to assure no other code is impacted. If we had it somewhere else, `TextEdit` would probably need to take non-pure-text input (i.e. streams of text blocks plus other stuff, much like `RichTextLabel`) and be able to report back events on non-text-input chunks; that would be a very big change as far as I see. Or we'd need to insert the circle as char, and then you have all the position problems coming.

blurymind commented:
This is fantastic!! Really hope to see it in godot master 👍

brainsick commented:
@poke1024 I definitely didn't mean to challenge this against the freeze.  I don't see why this wouldn't be viable for 3.0 yet.  Not that I have any say.  Thanks!  :)

poke1024 commented:
@brainsick No problem, I'm actually not that familiar with Godot API change rules, but AFAIK you could grab that signal from gdscript so it's probably part of the API.

akien-mga commented on scene/gui/text_edit.cpp:
```diff
@@ -30,6 +30,7 @@
 #include "text_edit.h"
 
 #include "message_queue.h"
+#include "modules/gdscript/gdscript_tokenizer.h"
```
GDScript is an optional module, we can't have a dependency on it in a core Control.

So this code should be moved somehow to the ScriptEditor, and guarded by `#ifdef GDSCRIPT_ENABLED`.

poke1024 commented:
@akien-mga All gdscript parsing and the color picker are now moved (back) to the ScriptEditor, I did a number of bug fixes, cleanups, and generalizations along the way.

This should make it relatively easy to embed other inline content in the future if needed.

As a simple test, I've added a triangular inline error indicator that shows the column when an error comes up. Just an experiment along the way.

<img width="687" alt="error" src="https://user-images.githubusercontent.com/11859538/34120091-c0fb491c-e424-11e7-9f41-0a12c9ba418c.png">

Paulb23 commented:
Whats the general consensus with this kind of inheritance / OOP should it use a new node type / signals instead?

poke1024 commented:
Currently `InlineProcessor` feels a bit like a hack. Note though that:
- `InlineProcessor::_parse_inline` gets called for every line while drawing
- the editor plugin system (e.g. `CollisionPolygonEditorPlugin`) use inheritance
- probably should not/need not be accessible from outside the tools

akien-mga commented:
Thanks a lot for your contribution!
We are now entering a strict release freeze for Godot 3.0 and will only consider major bug fixes. We won't merge new features and enhancements anymore until 3.0 is released.

Moving this PR to the 3.1 milestone, to be reviewed once the release freeze is lifted. It could eventually be cherry-picked for a future 3.0.1 maintenance release if it doesn't change the user-facing APIs and doesn't compromise the engine's stability.

vnen commented:
This would close #4882.

akien-mga commented:
This needs a rebase after the various TextEdit/CodeTextEditor abstractions done by @Paulb23 (#17923, #19225 and some others) which should also allow a cleaner implementation of this feature.

Maybe @Paulb23 could help review it?

poke1024 commented:
Rebased (which I don't see myself doing again) and cleaned up.

The implementation is basically the same as before, the new abstractions for the syntax coloring are too high level to make use fo them for rendering the additional inline elements.

There are still problems when wrapping is enabled (basically not yet implemented as this was not there before).

Paulb23 commented:
Tried this out locally, seems good for the most part.

It failed when the Colour is defined incorrectly:
```
var c = Color()
```
or when its defined over multiples lines e.g.
```
var c = Color(
    0,
    255
    0
)
```
In the first case, as its not valid code, will be erroring out at the tokenizer.
As for the second one, its probably the same reason, due it only inspecting one line a time, so the code won't be valid (most likely the same for wrapping).

It might be worth hiding the inline in these cases, rather then bothering the user with something they cannot interact with / displayed incorrectly.

As far as code goes, so long as the multi-inheritance is okay, I would move the tokenizer part further down something like:
`script->get_language()->parse_inlines()`

As a side thought, it might be worth looking at caching these in a similar way to `ColorRegions`.

vnen commented:
> It failed when the Colour is defined incorrectly:
>```
>var c = Color()
>```

This is valid code, I don't know why this would give any error.

Paulb23 commented:
> This is valid code, I don't know why this would give any error.

My bad, it doesn't error. As there are no arguments, it can't find the colour values so it doesn't display correctly / set it as editable.

reduz commented:
This branch has conflicts and some concerns of places where syntax wont work. Any progress on this so far?

poke1024 commented:
I'll take a look at this this weekend.

l1ebl committed:
TextEdit annotations: redesign

- Renamed "inline" to "annotation"
- Don't display partially parsable Colors
- Fix annotations with line wrapping
- Moved tokenizing to GDScriptLanguage
- Caching annotation vectors in TextEdit

Paulb23 reviewed (commented):
Its Looking good, There's a issue with the error marker sometimes not appearing, or appearing in the incorrect place:

![error_issue](https://user-images.githubusercontent.com/6584330/45321573-98716180-b53d-11e8-9677-761f5f2e6350.png)

But apart from that, only very minor code issues.

akien-mga commented:
Moving to 3.2 milestone as we're reaching beta stage for 3.1, and there are still changes needed in this PR.

Anutrix commented:
@poke1024 Rebase?

akien-mga commented:
This needs to be rebased indeed to be mergeable, but we haven't reached a clear consensus about the implementation.

The feature looks awesome, but I have some concerns over adding annotations to `ScriptLanguage` for this, and the extensive changes to `TextEdit` for a feature which is mostly relevant for editor tools.

I mentioned in other issues and on IRC that I find `TextEdit` to grow too big and beyond the scope it should have as a generic text input field for games.
I think we should aim to refactor this for 4.0 and have `TextEdit` be a lightweight text input control, and add a `CodeEdit` or similar that extends `TextEdit` and implements all code/tool-specific features like inline colors, but also syntax highlighting, indentation, breakpoint gutter, etc.

If we decide to do this, I would then hold off on this PR until such refactoring has been done, and then the PR could be redone to implement the feature via `CodeEdit` (or whatever its name would be). I'll open an issue about it.

akien-mga commented:
As mentioned in https://github.com/godotengine/godot/pull/14731#issuecomment-525653791, a correct implementation for this feature should likely wait for #31739 to be implemented. I will close this for now as it's not going to be merged as is.

This should likely be rediscussed in a proposal in the https://github.com/godotengine/godot-proposals/ repository, to see if/how we want to support this feature for the script editor (and shader editor?) while keeping the underlying controls manageable.
